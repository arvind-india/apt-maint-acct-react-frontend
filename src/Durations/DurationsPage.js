import React from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import { Table } from 'reactstrap'
import { Router } from 'react-router-dom'

import {
  MdAdd,
  MdVisibility,
  MdEdit,
  MdDelete
} from 'react-icons/lib/md' // material design icons

import {
  Button
} from 'reactstrap'

import { history } from '../_helpers'
import { PrivateRoute, FlashMessage } from '../_components'
import { durationActions as actions } from '../_actions'
import { default as detailsPage } from './DurationDetailsPage'

let url = '/durations'
let module = 'durations'

export class Durations extends React.Component {

  constructor(props) {
    super(props)
    this.handleDeleteModel = this.handleDeleteModel.bind(this)
  }

  componentDidMount() {
    this.props.getAll()
  }

  render() {
    const { durations, alert, authzn, trackHistory } = this.props
    let models = durations
    let hist = trackHistory?history:{}
    return (
      <div>
        <h3>Durations List</h3>
        {alert.message && <FlashMessage text={alert.message} color={alert.color} delay={2100}/>}
        {models.loading && <em>Loading models...}</em>}
        {models.error && <span className="text-danger">ERROR: {models.error}</span>}
        {models.items && authzn && this.showList() }
        <Router history={hist}>
          <div>
            <PrivateRoute path={`${url}/:id`} component={detailsPage} />
          </div>
        </Router>
      </div>
    )
  }

  showList(){
    const { durations } = this.props
    let models = durations
    return <Table>
      <thead>{ this.headerRow() }</thead>
      <tbody>
        {models.items.map((model, index) => this.bodyRow(model, index))}
      </tbody>
    </Table>
  }
  headerRow() {
    return <tr>
      <th>#</th>
      <th>Key</th>
      <th>Value</th>
      <th>Effective From</th>
      <th>Effective To</th>
      <th>Actions {this.addLink()}</th>
    </tr>
  }
  bodyRow(model, index) {
    return <tr key={model.id}>
      <th scope="row">{index+1}</th>
      <td>{model.key}</td>
      <td>{model.value}</td>
      <td>{model.effective_from}</td>
      <td>{model.effective_to}</td>
      { this.showActions(model) }
    </tr>
  }
  addLink() {
    const { authzn } = this.props
    return authzn.allowsAdd ?
      <Link to={{ pathname: `${url}/0`, state: this.newModel() }} title="Add"><MdAdd/></Link> :
      ''
  }

  newModel() {
    return {
      model: {
        id: 0,
        key:'',
        value: '',
        effective_from: '',
        effective_to: '',
        remarks: ''
      }
    }
  }

  showActions(model) {
    const { authzn } = this.props
    return <td>
            <Link
              to={{ pathname: `${url}/${model.id}`, state:{model: model} }}
              title={authzn.allowsEdit?"Edit":"View"}
            >{authzn.allowsEdit?<MdEdit/>:<MdVisibility/>}</Link>
            <Button
              color="link"
              title="Delete"
              onClick={() => this.handleDeleteModel(model.id)}
              hidden={!authzn.allowsDelete}
            ><MdDelete color="red"/></Button>
          </td>
  }

  handleDeleteModel(id) {
    if( window.confirm('Are you sure?') ) {
      this.props.delete(id)
      this.props.getAll() // get list after deletion of a model
    }
  }

}

function mapStateToProps(state) {
  const { durations, alert, authorizations } = state
  const authzn = authorizations[module]
  const trackHistory = true  // added for unit testing; snapshot to be precise
  return {
    durations,
    alert,
    authzn,
    trackHistory
  }
}

function mapDispatchToProps(dispatch) {
  return {
    getAll: () => {
      dispatch(actions.getAll())
    },
    delete: (id) => {
      dispatch(actions.delete(id))
    }
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Durations)
