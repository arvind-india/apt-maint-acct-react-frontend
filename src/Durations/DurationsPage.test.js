// DurationsPage.test.js

import React from 'react'
import { shallow } from 'enzyme'

import { Durations } from './DurationsPage'

const mockProps = {
  trackHistory: false,  // for unit testing history is NOT required as it has 'keys' that are dynamic
  durations: {
    loading: false,
    error: false,
    items: [
      {
        id: 1,
        key: 'guest',
        value: 'guest',
        effectiveFrom: '2018-03-31',
        effectiveTo: '2020-03-31',
        remarks: 'testing'
      },
      {
        id: 2,
        key: 'og1',
        value: 'g1',
        effectiveFrom: '2018-04-31',
        effectiveTo: '2020-04-31',
        remarks: 'testing'
      },
      {
        id: 3,
        key: 'tg1',
        value: 't1',
        effectiveFrom: '2018-05-31',
        effectiveTo: '2020-05-31',
        remarks: 'testing'
      }
    ]
  },
  alert: {message: "testingAlert"},
  authzn: {
    allowsAdd: true,
    allowsEdit: true,
    allowsDelete: true,
    condition: ''
  },
  getAll: jest.fn(),
  delete: jest.fn()
}

const component = shallow(<Durations {...mockProps} />)

describe('DurationsPage params test', () => {
  it('should render DurationsPage without crashing', () => {
    component
  })
  it('should render initial layout', () => {
    expect(component.getElements()).toMatchSnapshot()
  })
})

describe('DurationsPage rows test', () => {
  let numberOfMockDurations = mockProps.durations.items.length
  let tbody = component.find('tbody')
  let rows = tbody.find('tr')
  it('should have '+numberOfMockDurations+' rows', () => {
    expect(rows.length).toEqual(numberOfMockDurations)
  })
})
