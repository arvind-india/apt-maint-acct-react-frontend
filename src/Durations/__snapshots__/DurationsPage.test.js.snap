// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DurationsPage params test should render initial layout 1`] = `
Array [
  <div>
    <h3>
        Durations List
    </h3>
    <FlashMessage
        color={undefined}
        delay={2100}
        text="testingAlert"
    />
    <Table
        responsiveTag="div"
        tag="table"
    >
        <thead>
            <tr>
                <th>
                    #
                </th>
                <th>
                    Key
                </th>
                <th>
                    Value
                </th>
                <th>
                    Effective From
                </th>
                <th>
                    Effective To
                </th>
                <th>
                    Actions 
                    <Link
                        replace={false}
                        title="Add"
                        to={
                            Object {
                                "pathname": "/durations/0",
                                "state": Object {
                                  "model": Object {
                                    "effective_from": "",
                                    "effective_to": "",
                                    "id": 0,
                                    "key": "",
                                    "remarks": "",
                                    "value": "",
                                  },
                                },
                              }
                        }
                    >
                        <MdAdd />
                    </Link>
                </th>
            </tr>
        </thead>
        <tbody>
            <tr>
                <th
                    scope="row"
                >
                    1
                </th>
                <td>
                    guest
                </td>
                <td>
                    guest
                </td>
                <td />
                <td />
                <td>
                    <Link
                        replace={false}
                        title="Edit"
                        to={
                            Object {
                                "pathname": "/durations/1",
                                "state": Object {
                                  "model": Object {
                                    "effectiveFrom": "2018-03-31",
                                    "effectiveTo": "2020-03-31",
                                    "id": 1,
                                    "key": "guest",
                                    "remarks": "testing",
                                    "value": "guest",
                                  },
                                },
                              }
                        }
                    >
                        <MdEdit />
                    </Link>
                    <Button
                        color="link"
                        hidden={false}
                        onClick={[Function]}
                        tag="button"
                        title="Delete"
                    >
                        <MdDelete
                            color="red"
                        />
                    </Button>
                </td>
            </tr>
            <tr>
                <th
                    scope="row"
                >
                    2
                </th>
                <td>
                    og1
                </td>
                <td>
                    g1
                </td>
                <td />
                <td />
                <td>
                    <Link
                        replace={false}
                        title="Edit"
                        to={
                            Object {
                                "pathname": "/durations/2",
                                "state": Object {
                                  "model": Object {
                                    "effectiveFrom": "2018-04-31",
                                    "effectiveTo": "2020-04-31",
                                    "id": 2,
                                    "key": "og1",
                                    "remarks": "testing",
                                    "value": "g1",
                                  },
                                },
                              }
                        }
                    >
                        <MdEdit />
                    </Link>
                    <Button
                        color="link"
                        hidden={false}
                        onClick={[Function]}
                        tag="button"
                        title="Delete"
                    >
                        <MdDelete
                            color="red"
                        />
                    </Button>
                </td>
            </tr>
            <tr>
                <th
                    scope="row"
                >
                    3
                </th>
                <td>
                    tg1
                </td>
                <td>
                    t1
                </td>
                <td />
                <td />
                <td>
                    <Link
                        replace={false}
                        title="Edit"
                        to={
                            Object {
                                "pathname": "/durations/3",
                                "state": Object {
                                  "model": Object {
                                    "effectiveFrom": "2018-05-31",
                                    "effectiveTo": "2020-05-31",
                                    "id": 3,
                                    "key": "tg1",
                                    "remarks": "testing",
                                    "value": "t1",
                                  },
                                },
                              }
                        }
                    >
                        <MdEdit />
                    </Link>
                    <Button
                        color="link"
                        hidden={false}
                        onClick={[Function]}
                        tag="button"
                        title="Delete"
                    >
                        <MdDelete
                            color="red"
                        />
                    </Button>
                </td>
            </tr>
        </tbody>
    </Table>
    <Router
        history={Object {}}
    >
        <div>
            <PrivateRoute
                component={[Function]}
                path="/durations/:id"
            />
        </div>
    </Router>
</div>,
]
`;
